src/BufferType.cc

```diff
@@ -0,0 +1,210 @@
+  Py_buffer *view = (Py_buffer *)bufView;
+  PyBuffer_Release(view);
+  delete view;
+}
+
+/* static */
+JS::Scalar::Type BufferType::_getPyBufferType(Py_buffer *bufView) {
+  if (!bufView->format) { // If `format` is NULL, "B" (unsigned bytes) is assumed. https://docs.python.org/3.9/c-api/buffer.html#c.Py_buffer.format
+    return JS::Scalar::Uint8;
+  }
+  if (std::char_traits<char>::length(bufView->format) != 1) { // the type code should be a single character
+    return JS::Scalar::MaxTypedArrayViewType; // invalid
+  }
+
+  char typeCode = bufView->format[0];
+  // floating point types
+  if (typeCode == 'f') {
+    return JS::Scalar::Float32;
+  } else if (typeCode == 'd') {
+    return JS::Scalar::Float64;
+  }
+
+  // integer types
+  // We can't rely on the type codes alone since the typecodes are mapped to C types and would have different sizes on different architectures
+  //    see https://docs.python.org/3.9/library/array.html#module-array
+  //        https://github.com/python/cpython/blob/7cb3a44/Modules/arraymodule.c#L550-L570
+  // TODO (Tom Tang): refactor to something like `switch (typeCode) case 'Q': return [compile-time] intType<size: sizeof(long long), signed: true>`
+  bool isSigned = std::islower(typeCode); // e.g. 'b' for signed char, 'B' for unsigned char
+  uint8_t byteSize = bufView->itemsize;
+  switch (byteSize) {
+  case 1:
+    return isSigned ? JS::Scalar::Int8 : JS::Scalar::Uint8; // TODO (Tom Tang): Uint8Clamped
+  case 2:
+    return isSigned ? JS::Scalar::Int16 : JS::Scalar::Uint16;
+  case 4:
+    return isSigned ? JS::Scalar::Int32 : JS::Scalar::Uint32;
+  case 8:
+    return isSigned ? JS::Scalar::BigInt64 : JS::Scalar::BigUint64;
+  default:
+    return JS::Scalar::MaxTypedArrayViewType; // invalid byteSize
+  }
+}
+
+/* static */
+const char *BufferType::_toPyBufferFormatCode(JS::Scalar::Type subtype) {
+  // floating point types
+  if (subtype == JS::Scalar::Float32) {
+    return "f";
+  } else if (subtype == JS::Scalar::Float64) {
+    return "d";
+  }
+
+  // integer types
+  bool isSigned = JS::Scalar::isSignedIntType(subtype);
+  uint8_t byteSize = JS::Scalar::byteSize(subtype);
+  // Python `array` type codes are strictly mapped to basic C types (e.g., `int`), widths may vary on different architectures,
+  // but JS TypedArray uses fixed-width integer types (e.g., `uint32_t`)
+  switch (byteSize) {
+  case sizeof(char):
+    return isSigned ? "b" : "B";
+  case sizeof(short):
+    return isSigned ? "h" : "H";
+  case sizeof(int):
+    return isSigned ? "i" : "I";
+  // case sizeof(long): // compile error: duplicate case value
+  //                    // And this is usually where the bit widths on 32/64-bit systems don't agree,
+  //                    //    see https://en.wikipedia.org/wiki/64-bit_computing#64-bit_data_models
+  //   return isSigned ? "l" : "L";
+  case sizeof(long long):
+    return isSigned ? "q" : "Q";
+  default: // invalid
+    return "x"; // type code for pad bytes, no value
```

Comment:
We can't return a `nullptr` here because `"B"` (unsigned char) is assumed for null type code